### ISSUE TITLE
easy type upgrade for validate to better reflect uncertainty

### ISSUE THREAD

**Issue by dgreene1** [jaredpalmer/formik#1590]
### Current Behavior

When using `withFormik`, the FormValues won't necessarily be truthy at runtime; however, the types imply that they will. So I'd like to make a PR to have the types match reality.

### Expected behavior

When calling `withFormik<OutProps, FormValues>(valuesToTest)` the type of `valuesToTest`  should be `Partial<FormValues>`

### Reproducible example

I'll provide a tiny snippet of the code from the documentation to show that the general idea is test each property valuesToTest:
```
// Wrap our form with the using withFormik HoC
const MyForm = withFormik<{ aId: string }, { aId: string, bId: string, cId: string }>({

    // Add a custom validation function (this can be async too!)
    validate: (values) => {
        let errors: FormikErrors = {};
        if (!values.bId) {
            errors.bId= 'Required';
        } else if (!isValidEmail(values.email)) {
        return errors;
    },

    handleSubmit: values => {
        // do submitting things
    },
})(InnerForm);
```

### Suggested solution(s)

Long story short, if you're okay with me making an update to wrap the type in `Partial<T>` I would be more than happy to make that PR. I suspect that this will not be a breaking change for anyone  because most people will be checking the truthiness of the form values inside of the `validate` function anyway.

**Comment by johnrom** [comment]
This isn't necessarily correct. Partial and "falsiness" have completely different meaning in TypeScript.

```
interface PartialFormValues {
    value?: string;
}
const formValues: PartialFormValues = {}; // valid

interface UninitializedFormValues {
    value: string | undefined;
}
const formValues: UninitializedFormValues = {}; // invalid
const formValues: UninitializedFormValues = { value: undefined }; // valid
const formValues: Partial<UninitializedFormValues> = {}; // valid but not the same type as you defined for the form

interface InitializedFormValues {
    value: string;
}
const formValues: InitializedFormValues = {}; // invalid
const formValues: InitializedFormValues = { value: undefined }; // invalid
const formValues: Partial<InitializedFormValues> = {}; // valid but still not the same type as you defined for the form
const formValues: InitializedFormValues = { value: '' }; // valid, though falsy
```

In your example, you seem to be questioning truthiness, when Partial doesn't actual concern truthiness, but existence itself. The form should be composed of values (whether truthy or not) wherein all the non-optional properties in FormValues exist. If you mark an option as "optional" like in PartialFormValues, that's OK, but Formik isn't going to do that work for you, or it will break the expectations you've set up by using TypeScript.

**Comment by dgreene1** [comment]
That was a strangely pedantic response @johnrom. I understand the difference between existence and the literal value of undefined. But the point is that `Partial<T>` is an easy way to represent the fact that upon reading the validate callback, any of the properties on the form value may not be truthy. The runtime values do not match the types, simple as that. If I am wrong about the runtime behavior, then feel free to close the PR.

**Comment by johnrom** [comment]
@dgreene1 I'm a bit confused. No offense intended by writing up that information. I don't usually use withFormik so I'm probably misunderstanding your question. With `<Formik />`, `initialValues={{}}` must always match the signature of your type, so `Partial<T>` would violate that expectation. If I expected a state where, during validation, a value might not exist, I would make that clear in the type that I defined initially. Maybe `withFormik` has a different meaning than I expect it to. In such an event, someone else may be more suited to respond to this issue.

**Comment by dgreene1** [comment]
It’s fine. I just wasn’t expecting a lesson in something I already understood.

But I think the root of our miscommunication might stem from the fact that the documentation shows to set the initial values to an empty string, which explains why the validate function in that [same example](https://jaredpalmer.com/formik/docs/guides/typescript) is using a truthiness check since that also catches empty string. I’m not a big fan of initializing strings to empty string since I feel it subverts the type system. Like at runtime this is false:
`!!””`
But at compilation time “” is of type string.

So perhaps I’m pointing a larger issue with Formik which is that it doesn’t appear possible to explicitly represent the scenario where you want a form to have no initial values but where it must have certain values upon completion of the form.

So something like this would be awesome:
```
withFormikStrict<TInitialValues, TFinalValues>(...)
```
Because then validate would send TInitialValues. This would enable self-documenting code like:
```
type InitialValues = undefined;
const MyForm = withFormik<

**Comment by johnrom** [comment]
Since an issue is for posterity, I try to provide as much context as possible!

In terms of your request itself, I think someone with more authority will have to speak to that.

**Comment by jaredpalmer** [comment]
I disagree about subverting the type system as an empty string can be a legitimate value. It is thus not a partial.

**Comment by dgreene1** [comment]
@jaredpalmer, it might be a legitimate value, but it’s not a value that satisfies the required check, right?

So is there a way to use Formik so that you give a form no initial values?

Because when I fill out a doctor’s form, there are no pre fill fields. Every value is undefined.

That’s why I was recommending an approach that lets the dev set the undefined initial state and the desired final state. Then the TypeScript types would guide future devs on what validation checks they should be writing. Wouldn’t that be cool?

Would you accept a PR for something like that?

**Comment by jaredpalmer** [comment]
If you do not properly initialize a field with an initial value, React will throw an error saying that you changed an input from being uncontrolled to controlled. So no. initialValues is required.

**Comment by jaredpalmer** [comment]
I would not accept a PR for this. The text input values of your doctor's form should be empty string and are not partial. FWIW this is how React initializes empty input values (see here for official documentation on forms: https://reactjs.org/docs/forms.html#controlled-components). Put differently....

Here is the name form from the React docs on forms:

```jsx
class NameForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {value: ''};

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    this.setState({value: event.target.value});
  }

  handleSubmit(event) {
    alert('A name was submitted: ' + this.state.value);
    event.preventDefault();
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Name:
          <input type="text" value={this.state.value} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Submit" />
      </form>
    );
  }
}
```
in typescript you would write this as follows:

```tsx
interface State {
  value: string
}
class NameForm extends React.Component<{}, State> {
 // ...
}
```

You would not call type this as `Partial<State>` since it is not Partial. Formik mimics React exactly here.

**Comment by dgreene1** [comment]
Thank you for explaining that @jaredpalmer. I did not know that about uncontrolled forms. Additionally, you helped me to understand that I should test for empty (and therefore whitespace) in my validations. This was helpful. :)
